deathcause.lua

```diff
@@ -0,0 +1,212 @@
+    'experienced death by QUIT',    -- QUIT
+    'experienced death by ABANDON',    -- ABANDON
+    'suffered heat stroke',        -- HEAT
+    'died of hypothermia',            -- COLD
+    'experienced death by SPIKE',    -- SPIKE
+    'experienced death by ENCASE_LAVA',    -- ENCASE_LAVA
+    'experienced death by ENCASE_MAGMA',    -- ENCASE_MAGMA
+    'was preserved in ice',        -- ENCASE_ICE
+    'became headless',                -- BEHEAD
+    'was crucified',                -- CRUCIFY
+    'experienced death by BURY_ALIVE',    -- BURY_ALIVE
+    'experienced death by DROWN_ALT',    -- DROWN_ALT
+    'experienced death by BURN_ALIVE',    -- BURN_ALIVE
+    'experienced death by FEED_TO_BEASTS',    -- FEED_TO_BEASTS
+    'experienced death by HACK_TO_PIECES',    -- HACK_TO_PIECES
+    'choked on air',                -- LEAVE_OUT_IN_AIR
+    'experienced death by BOIL',    -- BOIL
+    'melted',                        -- MELT
+    'experienced death by CONDENSE',    -- CONDENSE
+    'experienced death by SOLIDIFY',    -- SOLIDIFY
+    'succumbed to infection',        -- INFECTION
+    "s ghost was put to rest with a memorial",    -- MEMORIALIZE
+    'scared to death',                -- SCARE
+    'experienced death by DARKNESS',    -- DARKNESS
+    'experienced death by COLLAPSE',    -- COLLAPSE
+    'was drained of blood',            -- DRAIN_BLOOD
+    'was slaughtered',                -- SLAUGHTER
+    'became roadkill',                -- VEHICLE
+    'killed by a falling object',    -- FALLING_OBJECT
+}
+
+-- Creates a table of all items at the given location optionally matching a given item type
+function item_find(pos_x, pos_y, pos_z, item_type)
+    local items = {}
+    for _, item in ipairs(df.global.world.items.all) do
+        if item.pos.x == pos_x and item.pos.y == pos_y and item.pos.z == pos_z then
+            if not item_type or item:getType() == item_type then
+                table.insert(items, item)
+            end
+        end
+    end
+    return items
+end
+
+-- Finds a unit with the given id or nil if no unit is found
+function unit_find(unit_id)
+    return utils.binsearch(df.global.world.units.all, unit_id, 'id')
+end
+
+function getRace(race_id)
+    return df.global.world.raws.creatures.all[race_id]
+end
+
+function getRaceNameSingular(race_id)
+    return getRace(race_id).name[0]
+end
+
+function getDeathStringFromCause(cause)
+    if cause == -1 then
+        return "none"
+    else
+        return DEATH_TYPES[cause]
+    end
+end
+
+function display_death_unit(unit)
+    local str = string.format("The %s", getRaceNameSingular(unit.race))
+    if unit.name.has_name then
+        str = str .. string.format(" %s", dfhack.TranslateName(unit.name))
+    end
+
+    if not unit.flags2.killed and not unit.flags3.ghostly then
+        str = str .. " is not dead yet!"
+        print(str)
+        return
+    end
+
+    str = str .. string.format("%s", getDeathStringFromCause(unit.counters.death_cause))
```

Comment:
what is the format adding here? if you are just protecting against `nil` concatenation, `tostring()` might be cleaner.

as an aside that I didn't want to make a comment for elsewhere, there is an alternate syntax for calling string functions that's a little shorter: `formatstring:format(values)`. note that string literals need to be surrounded by parentheses, e.g. `('%s'):format(getDeathStringFromCause(unit.counters.death_cause))`

you might already know about it, and either format is fine. I just wanted to make sure you were aware of your options.

Reply:
I think this format was just left in while I was in the process of figuring stuff out and didn't get cleaned up.

I was not aware there were other forms for format though, thank you!

Reply:
This format was actually wrong, and it needed a space to be added before the string. Side-effect of refactoring.

Reply:
> I was not aware there were other forms for format though, thank you!

not just for format, any string function can use the `obj:classFunction()` syntax. DFHack adds a few extra string functions to the standard library as well: https://docs.dfhack.org/en/stable/docs/dev/Lua%20API.html#string-class-extensions